Jetty 12.1.6 HttpFields from(HttpField...) does not provide opportunity to set compliance listener

**Jetty version(s)**
<!--[Jetty 9.x is now at End of Community Support](https://github.com/jetty/jetty.project/issues/7958) -->
Jetty 12.1.6

**Jetty Environment**
<!-- Applicable for jetty-12 only, choose: core, ee8, ee9, ee10, ee11 -->
EE11

**HTTP version**
<!-- HTTP 1.0, 1.1, 2 or 3 -->
HTTP/1.1

**Java version/vendor** `(use: java -version)`
openjdk version "25.0.2" 2026-01-20 LTS
OpenJDK Runtime Environment Zulu25.32+22-SA (build 25.0.2+10-LTS)
OpenJDK 64-Bit Server VM Zulu25.32+22-SA (build 25.0.2+10-LTS, mixed mode, sharing)

**OS type/version**
Windows 11 x64

**Description**
<!-- Do not report security issues here! See [Jetty Security Reports](https://jetty.org/security.html) -->
The HTTP compliance enhancements in https://github.com/jetty/jetty.project/issues/14397 now provide a method of setting custom HTTP compliance and listeners to HTTP header parsing. However, we observe that a previous 12.1.5 workflow provides no clear opportunity to set a listener, resulting in a NPE.

When attempting to use the HttpFields from factory to build the object:

```
    HttpFields fields = HttpFields.from(new HttpField(HttpHeader.ACCEPT_LANGUAGE, acceptLanguageHeader));
    List<String> acceptable = fields.getQualityCSV(HttpHeader.ACCEPT_LANGUAGE);
```

we observe:

```
  Caused by: org.eclipse.jetty.websocket.core.exception.WebSocketException: WebSocket OPEN method error: Cannot invoke "java.util.function.Supplier.get()" because "this._listenerSupplier" is null
        ... 36 more
  Caused by: java.lang.NullPointerException: Cannot invoke "java.util.function.Supplier.get()" because "this._listenerSupplier" is null
        at org.eclipse.jetty.http@12.1.6/org.eclipse.jetty.http.ImmutableHttpFields.newQuotedQualityCSV(ImmutableHttpFields.java:167)
        at org.eclipse.jetty.http@12.1.6/org.eclipse.jetty.http.HttpFields.getQualityCSV(HttpFields.java:787)
        at org.eclipse.jetty.http@12.1.6/org.eclipse.jetty.http.HttpFields.getQualityCSV(HttpFields.java:763)
```

It appears that the Immutable factory sequence:

```
    /**
     * <p>Returns an immutable {@link HttpFields} instance containing the given
     * {@link HttpField}s.</p>
     *
     * @param fields the {@link HttpField}s to be contained in the returned instance
     * @return a new immutable {@link HttpFields} instance with the given {@link HttpField}s
     */
    static HttpFields from(HttpField... fields)
    {
        return new org.eclipse.jetty.http.ImmutableHttpFields(fields);
    }

...

    /**
     * Initialize HttpFields from copy.
     *
     * @param fields the fields to copy data from
     */
    protected ImmutableHttpFields(HttpField[] fields)
    {
        this(fields, fields.length);
    }

    protected ImmutableHttpFields(HttpField[] fields, int size)
    {
        this(null, null, fields, size);
    }

    ImmutableHttpFields(HttpCompliance httpCompliance, Supplier<ComplianceViolation.Listener> listenerSupplier, HttpField[] fields, int size)
    {
        _httpCompliance = httpCompliance;
        _listenerSupplier = listenerSupplier;
        _fields = fields;
        _size = size;
    }
```

Provides no workflow for setting '_listenerSupplier ' which later produces NPE at:

```
  @Override
    public QuotedCSV newQuotedCSV(boolean keepQuotes)
    {
        return new QuotedCSV(_httpCompliance, _listenerSupplier.get(), keepQuotes);
    }
```

We have worked around this issue with the following alternative:

```
    HttpFields.Mutable headers = HttpFields.build();
    headers.put(new HttpField(HttpHeader.ACCEPT_LANGUAGE, acceptLanguageHeader));
    HttpFields fields = headers.asImmutable();
    List<String> acceptable = fields.getQualityCSV(HttpHeader.ACCEPT_LANGUAGE);
```

as this uses, and copies to the immutable instance, the default listener:

```
  protected MutableHttpFields(int capacity)
    {
        _httpCompliance = null;
        _listenerSupplier = () -> null;
        _fields = new HttpField[capacity];
    }
```

It seems as though using the default listener () -> null might be an appropriate action for the HttpFields.from() workflow.

**How to reproduce?**
```
    String acceptLanguageHeader = completedUpgradeRequest.getHeader(HttpHeader.ACCEPT_LANGUAGE.toString());

    ...

    HttpFields fields = HttpFields.from(new HttpField(HttpHeader.ACCEPT_LANGUAGE, acceptLanguageHeader));
    List<String> acceptable = fields.getQualityCSV(HttpHeader.ACCEPT_LANGUAGE);
```

> **Jetty version(s)**
> 
> Jetty 12.1.6
> 
> **Jetty Environment**
> 
> EE11
> 
> **HTTP version**
> 
> HTTP/1.1
> 
> **Java version/vendor** `(use: java -version)` openjdk version "25.0.2" 2026-01-20 LTS OpenJDK Runtime Environment Zulu25.32+22-SA (build 25.0.2+10-LTS) OpenJDK 64-Bit Server VM Zulu25.32+22-SA (build 25.0.2+10-LTS, mixed mode, sharing)
> 
> **OS type/version** Windows 11 x64
> 
> **Description**
> 
> The HTTP compliance enhancements in [#14397](https://github.com/jetty/jetty.project/issues/14397) now provide a method of setting custom HTTP compliance and listeners to HTTP header parsing. However, we observe that a previous 12.1.5 workflow provides no clear opportunity to set a listener, resulting in a NPE.
> 
> When attempting to use the HttpFields from factory to build the object:
> 
> ```
>     HttpFields fields = HttpFields.from(new HttpField(HttpHeader.ACCEPT_LANGUAGE, acceptLanguageHeader));
>     List<String> acceptable = fields.getQualityCSV(HttpHeader.ACCEPT_LANGUAGE);
> ```
> 
> we observe:
> 
> ```
>   Caused by: org.eclipse.jetty.websocket.core.exception.WebSocketException: WebSocket OPEN method error: Cannot invoke "java.util.function.Supplier.get()" because "this._listenerSupplier" is null
>         ... 36 more
>   Caused by: java.lang.NullPointerException: Cannot invoke "java.util.function.Supplier.get()" because "this._listenerSupplier" is null
>         at org.eclipse.jetty.http@12.1.6/org.eclipse.jetty.http.ImmutableHttpFields.newQuotedQualityCSV(ImmutableHttpFields.java:167)
>         at org.eclipse.jetty.http@12.1.6/org.eclipse.jetty.http.HttpFields.getQualityCSV(HttpFields.java:787)
>         at org.eclipse.jetty.http@12.1.6/org.eclipse.jetty.http.HttpFields.getQualityCSV(HttpFields.java:763)
> ```
> 
> It appears that the Immutable factory sequence:
> 
> ```
>     /**
>      * <p>Returns an immutable {@link HttpFields} instance containing the given
>      * {@link HttpField}s.</p>
>      *
>      * @param fields the {@link HttpField}s to be contained in the returned instance
>      * @return a new immutable {@link HttpFields} instance with the given {@link HttpField}s
>      */
>     static HttpFields from(HttpField... fields)
>     {
>         return new org.eclipse.jetty.http.ImmutableHttpFields(fields);
>     }
> 
> ...
> 
>     /**
>      * Initialize HttpFields from copy.
>      *
>      * @param fields the fields to copy data from
>      */
>     protected ImmutableHttpFields(HttpField[] fields)
>     {
>         this(fields, fields.length);
>     }
> 
>     protected ImmutableHttpFields(HttpField[] fields, int size)
>     {
>         this(null, null, fields, size);
>     }
> 
>     ImmutableHttpFields(HttpCompliance httpCompliance, Supplier<ComplianceViolation.Listener> listenerSupplier, HttpField[] fields, int size)
>     {
>         _httpCompliance = httpCompliance;
>         _listenerSupplier = listenerSupplier;
>         _fields = fields;
>         _size = size;
>     }
> ```
> 
> Provides no workflow for setting '_listenerSupplier ' which later produces NPE at:
> 
> ```
>   @Override
>     public QuotedCSV newQuotedCSV(boolean keepQuotes)
>     {
>         return new QuotedCSV(_httpCompliance, _listenerSupplier.get(), keepQuotes);
>     }
> ```
> 
> We have worked around this issue with the following alternative:
> 
> ```
>     HttpFields.Mutable headers = HttpFields.build();
>     headers.put(new HttpField(HttpHeader.ACCEPT_LANGUAGE, acceptLanguageHeader));
>     HttpFields fields = headers.asImmutable();
>     List<String> acceptable = fields.getQualityCSV(HttpHeader.ACCEPT_LANGUAGE);
> ```
> 
> as this uses, and copies to the immutable instance, the default listener:
> 
> ```
>   protected MutableHttpFields(int capacity)
>     {
>         _httpCompliance = null;
>         _listenerSupplier = () -> null;
>         _fields = new HttpField[capacity];
>     }
> ```
> 
> It seems as though using the default listener () -> null might be an appropriate action for the HttpFields.from() workflow.
> 
> **How to reproduce?**
> 
> ```
>     String acceptLanguageHeader = completedUpgradeRequest.getHeader(HttpHeader.ACCEPT_LANGUAGE.toString());
> 
>     ...
> 
>     HttpFields fields = HttpFields.from(new HttpField(HttpHeader.ACCEPT_LANGUAGE, acceptLanguageHeader));
>     List<String> acceptable = fields.getQualityCSV(HttpHeader.ACCEPT_LANGUAGE);
> ```

--- a/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java
+++ b/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java
@@ -178,6 +178,8 @@ public interface HttpFields extends Iterable<HttpField>, Supplier<HttpFields>
     /**
      * <p>Returns an immutable {@link HttpFields} instance containing the given
      * {@link HttpField}s.</p>
+     * <p>The returned instance uses default HTTP compliance behavior (no explicit
+     * compliance mode and no-op compliance violation listener supplier).</p>
      *
      * @param fields the {@link HttpField}s to be contained in the returned instance
      * @return a new immutable {@link HttpFields} instance with the given {@link HttpField}s

--- a/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/ImmutableHttpFields.java
+++ b/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/ImmutableHttpFields.java
@@ -43,13 +43,13 @@ class ImmutableHttpFields implements HttpFields
 
     protected ImmutableHttpFields(HttpField[] fields, int size)
     {
-        this(null, () -> null, fields, size);
+        this(null, null, fields, size);
     }
 
     ImmutableHttpFields(HttpCompliance httpCompliance, Supplier<ComplianceViolation.Listener> listenerSupplier, HttpField[] fields, int size)
     {
         _httpCompliance = httpCompliance;
-        _listenerSupplier = listenerSupplier;
+        _listenerSupplier = Objects.requireNonNullElse(listenerSupplier, () -> null);
         _fields = fields;
         _size = size;
     }

--- a/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/ImmutableHttpFields.java
+++ b/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/ImmutableHttpFields.java
@@ -43,7 +43,7 @@ class ImmutableHttpFields implements HttpFields
 
     protected ImmutableHttpFields(HttpField[] fields, int size)
     {
-        this(null, null, fields, size);
+        this(null, () -> null, fields, size);
     }
 
     ImmutableHttpFields(HttpCompliance httpCompliance, Supplier<ComplianceViolation.Listener> listenerSupplier, HttpField[] fields, int size)

--- a/jetty-core/jetty-http/src/test/java/org/eclipse/jetty/http/HttpFieldsTest.java
+++ b/jetty-core/jetty-http/src/test/java/org/eclipse/jetty/http/HttpFieldsTest.java
@@ -44,6 +44,7 @@ import static org.hamcrest.Matchers.containsString;
 import static org.hamcrest.Matchers.instanceOf;
 import static org.hamcrest.Matchers.is;
 import static org.hamcrest.Matchers.nullValue;
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertFalse;
 import static org.junit.jupiter.api.Assertions.assertNull;
@@ -830,6 +831,24 @@ public class HttpFieldsTest
         assertEquals(HttpField.getValueParameters(list.get(3), null), "two");
         assertEquals(HttpField.getValueParameters(list.get(4), null), "three");
         assertEquals(HttpField.getValueParameters(list.get(5), null), "four");
+    }
+
+    private static Stream<Arguments> immutableFieldsForQualityCSV()
+    {
+        String value = "en-US;q=0.8,en;q=0.6";
+        return Stream.of(
+            Arguments.of("from", HttpFields.from(new HttpField(HttpHeader.ACCEPT_LANGUAGE, value))),
+            Arguments.of("buildAsImmutable", HttpFields.build().add(HttpHeader.ACCEPT_LANGUAGE, value).asImmutable())
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("immutableFieldsForQualityCSV")
+    public void testImmutableGetQualityCSV(String scenario, HttpFields fields)
+    {
+        List<String> list = assertDoesNotThrow(() -> fields.getQualityCSV(HttpHeader.ACCEPT_LANGUAGE), scenario);
+        assertEquals(HttpField.getValueParameters(list.get(0), null), "en-US");
+        assertEquals(HttpField.getValueParameters(list.get(1), null), "en");
     }
 
     @Test

--- a/jetty-core/jetty-http/src/test/java/org/eclipse/jetty/http/HttpFieldsTest.java
+++ b/jetty-core/jetty-http/src/test/java/org/eclipse/jetty/http/HttpFieldsTest.java
@@ -833,20 +833,11 @@ public class HttpFieldsTest
         assertEquals(HttpField.getValueParameters(list.get(5), null), "four");
     }
 
-    private static Stream<Arguments> immutableFieldsForQualityCSV()
+    @Test
+    public void testFromGetQualityCSV()
     {
-        String value = "en-US;q=0.8,en;q=0.6";
-        return Stream.of(
-            Arguments.of("from", HttpFields.from(new HttpField(HttpHeader.ACCEPT_LANGUAGE, value))),
-            Arguments.of("buildAsImmutable", HttpFields.build().add(HttpHeader.ACCEPT_LANGUAGE, value).asImmutable())
-        );
-    }
-
-    @ParameterizedTest
-    @MethodSource("immutableFieldsForQualityCSV")
-    public void testImmutableGetQualityCSV(String scenario, HttpFields fields)
-    {
-        List<String> list = assertDoesNotThrow(() -> fields.getQualityCSV(HttpHeader.ACCEPT_LANGUAGE), scenario);
+        HttpFields fields = HttpFields.from(new HttpField(HttpHeader.ACCEPT_LANGUAGE, "en-US;q=0.8,en;q=0.6"));
+        List<String> list = assertDoesNotThrow(() -> fields.getQualityCSV(HttpHeader.ACCEPT_LANGUAGE));
         assertEquals(HttpField.getValueParameters(list.get(0), null), "en-US");
         assertEquals(HttpField.getValueParameters(list.get(1), null), "en");
     }

